_posts/2023-08-30-spiffe.md

```diff
@@ -0,0 +1,364 @@
+
+We require a role binding named 'file-read-app' for our application.
+Create a new file named 'rolebinding.yaml' with the following content.
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: file-read-app
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: file-read-app
+subjects:
+  - kind: ServiceAccount
+    name: file-read-app
+```
+
+Deploy role binding.
+
+```bash
+kubectl apply -f rolebinding.yaml
+```
+
+### Deploy application
+
+Create a new file named 'deployment.yaml' with the following content.
+If necessary, update the environment variables.
+Fill in the three AWS resources with the one you have generated.
+
+SPIFFE will mount files in the volume we created here, SPIFFE would create a certificate to communicat with AWS. 
+After the certificate is genereated SPIFFE will use this to retrieve aws credentials and also mount this.
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: demo-spiffe
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: demospiffe
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: demospiffe
+    spec:
+      serviceAccount: file-read-app
+      containers:
+        - name: demospiffe
+          image: nicholas95/spiffedemo:dev-1
+          imagePullPolicy: Always
+          env:
+            - name: AWS_REGION
+              value: eu-west-1
+            - name: AWS_BUCKET
+              value: demospiffe
+          volumeMounts:
+            - mountPath: /root/.aws
+              name: spiffe
+      volumes:
+        - csi:
+            driver: spiffe.csi.cert-manager.io
+            readOnly: true
+            volumeAttributes:
+              aws.spiffe.csi.cert-manager.io/enable: "true"
+              aws.spiffe.csi.cert-manager.io/role: arn:aws:iam::930970667460:role/nicholas_meyers_home_k8s_read
+              aws.spiffe.csi.cert-manager.io/trust-anchor: arn:aws:rolesanywhere:eu-west-1:930970667460:trust-anchor/1bf58d4e-35c6-465c-a05c-1bdcce05054b
+              aws.spiffe.csi.cert-manager.io/trust-profile: arn:aws:rolesanywhere:eu-west-1:930970667460:profile/fa6c37f9-f64a-4bdf-b0f6-c39ec5282aa4
+          name: spiffe
+```
+
+Deploy application.
+
+```bash
+kubectl apply -f deployment.yaml
+```
+### Test our application
+
+When your pod is started, you can open a shell into it using the following command and install curl. 
+Once curl is installed, you can send a GET request to '/file/hello.txt', the response will be the content of the file in the S3 bucket.
+
+```bash
+kubectl exec -it <pod-name> -- /bin/sh
+apk add curl
+curl http://localhost:8080/file/hello.txt
+exit
+```
+
+We've also included a file watcher to monitor the renewal of credentials and certificates.
+
+<img src="{{ '/img/2023-08-30-spiffe/spiffe-5.png' | prepend: site.baseurl }}" alt="Filewatcher logging" class="image fit" style="margin:0px auto; max-width:100%">
+
+
+
+
+## Conclusion
```

Comment:
Please include a comparison with standard IRSA (https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) (aka workload identity) and this. (E.g. explain that you'd could use this approach to allow a local minikube / docker desktop installation of k8s to connect to AWS)

Explain that it can also be used outside of the K8s context. Compare it to what OpenIDConnect has done for user based SSO. 

Store all the example code in repository on your account and link it. This will improve "I want to test this" for users.

Reply:
repository mentioned in #extra